### ISSUE TITLE
memoize fails to handle code sensitive to newlines (such as inline tables well in tikz)

### ISSUE THREAD

**Issue by JasonGross** [sasozivanovic/memoize#19]
This code works only if I remove `\usepackage{memoize}`:
```
\documentclass{article}
\usepackage{pgfplots}
\pgfplotsset{compat=newest}
\usepackage{memoize}
\begin{document}
\begin{tikzpicture}
\begin{axis}
\addplot [semithick, black]
table {%
0 0
0.0752933174371719 0.0355731099843979
};
\end{axis}
\end{tikzpicture}
\end{document}
```
As it stands, it gives
[test2-memo-simple.log](https://github.com/sasozivanovic/memoize/files/15279935/test2-memo-simple.log)
```
! Package pgfplots Error: Could not read table file '"0 0 0.0752933174371719 0.
0355731099843979 "' in 'search path=.'. In case you intended to provide inline
data: maybe TeX screwed up your end-of-lines? Try `row sep=crcr' and terminate
your lines with `\\' (refer to the pgfplotstable manual for details).

See the pgfplots package documentation for explanation.
Type  H <return>  for immediate help.
 ...

l.14 \end{tikzpicture}
```
The `test2-memo-simple.memo.dir/7332D1E590C41D3A0D2C17379ABA8876.memo` file has
```latex
\mmzMemo
\global \mmzContextExtra {}%
%
\mmzSource
\begin {tikzpicture} \begin {axis} \addplot [semithick, black] table {0 0 0.0752933174371719 0.0355731099843979 }; \end {axis} \end {tikzpicture}
```
so I guess the issue is that newlines actually matter when externing?

**Comment by sasozivanovic** [comment]
I'm not familiar with inline tables, and I'm away from my computer right now, so I can't test this, but I believe that ˋverbatimˋ should help.

Try writing either ˋ\mmznext{verbatim}ˋ before ˋ\begin{tikzpicture}ˋ, or --- if all your TikZ pictures contain these tables --- ˋ\mmzsez{auto={tikzpicture}{verbatim}ˋ.

**Comment by JasonGross** [comment]
That sounds like it should fix it, thanks!  Is there any reason verbatim isn't the default?

**Comment by cfr42** [comment]
For what it is worth, `verbatim` works fine here. It works even if memoize is loaded _after_ pgfplots, which surprised me. 

Alternatively, you can use `row sep=\\`, as suggested on page 49 of the pgfplots manual, and separate the rows by `\\`. Either produce an extern successfully.

**Comment by cfr42** [comment]
> That sounds like it should fix it, thanks! Is there any reason verbatim isn't the default?

You lose catcode changes. Without verbatim, if I change the language from French to UK English, say, with Babel, the contents of a tikzpicture which includes a colon will get recompiled. With verbatim, it won't. 

[I'm sure there are better examples ....]

**Comment by sasozivanovic** [comment]
I was considering the idea to make `verbatim` default, but decided against it, mainly as a precaution. "Thou shalt not change catcodes in vain" — while Don Knuth never actually wrote this, he could have ;-)

cfr's example is a good one. Here's another:

```
\documentclass{article}
\usepackage{memoize}
\newcommand\cmd[1][42]{``#1''}
\mmzset{auto=\cmd{args={o}, memoize, verbatim}}
\begin{document}
bla \cmd % a comment
 [24]
bla 
\end{document}
```
In a nutshell, collecting arguments is hard work, and verbatim mode complicates matters even more. If you want to learn more on the borderline cases where Memoize (actually, CollArgs) fails, see subsection `Mode transition limitations` at the end of 5.6.3 (the reference section of CollArgs).

All that said, it is easy enough to have verbatim mode the default, simply write `\mmzset{verbatim}` either in the preamble or into the configuration file `memoize.cfg`. I'd be interested to hear if it causes any trouble in the real world.

**Comment by JasonGross** [comment]
> You lose catcode changes.

This should be possible to fix, at least in theory.  When setting tokens to be other, you could create macros `\csname \detokenize{<token>}@orig catcode\endcsname`.  Then when collecting tokens, check if this macro exists for each token, and if it does, emit the expansion of the macro instead of the token.  Alternatively, you could instead set catcodes to active rather than other, and define each token to expand to the version of itself with the catcode it originally had, but then you have to jump through some more complicated hoops to handle `\def\foo{\bar}\expandafter\collectargs\expandafter{\foo}`, since, IIRC, catcodes do not get reprocessed in this case (though it's possible I'm misremembering, it's been a while since I've dug into how TeX's mouth works).

I agree that verbatim mode can have unintended consequences and the above suggestion may be overkill, but I think it may be worth adding a mode that is equivalent to `\obeyspaces\obeylines` (without changing any other catcodes) and make this mode the default.

**Comment by JasonGross** [comment]
Could we reopen this issue?  I'm getting the very weird error
```
! Argument of \begin{tikzpicture} has an extra }.
<inserted text> 
                \par 
l.147 ...liationsAndNotice{\icmlEqualContribution}
                                                   % otherwise use the stand...
```
with verbatim mode in the icml2024 style, and would appreciate a mode of memoization that is not fully verbatim but still obeys newlines (and possibly spaces as well)

**Comment by cfr42** [comment]
@JasonGross Do you have an example?

**Comment by sasozivanovic** [comment]
> Could we reopen this issue?

Of course!

I'll look into this asap. As asked by @cfr42, an example would be useful.

**Comment by sasozivanovic** [comment]
I found the [ICML style](https://icml.cc/Conferences/2024/AuthorInstructions) on the web, and tracked the problem to package/environment `algorithmic`. Newlines / verbatim mode don't seem to be the issue here, but rather that this environment produces a vertical box. The solution is then `capture=vbox`.

To compile the `example_paper.tex` with Memoize, I included the following at the top of the file.

```
\usepackage{memoize}
\mmzset{
  auto={algorithmic}{memoize,capture=vbox},
}
```

Would you have any suggestions as to where to hint at this issue in the tutorial? Perhaps the verbatim section (2.9, which will receive an update in accord with issue #17) is a good place?

**Comment by JasonGross** [comment]
Wow, thanks for tracking this down!  I'm a bit confused, still: my full document works without `[verbatim]` but gives this error with `[verbatim]`; is this the same behavior you see with the example file? 

In any case, the verbatim section seems a fine place to hint this.

**Comment by sasozivanovic** [comment]
> Wow, thanks for tracking this down! 

To be honest, I was stumped for a while. ;-)
So having this as a reminder in the tutorial will be useful for me, too!

> I'm a bit confused, still: my full document works without `[verbatim]` but gives this error with `[verbatim]`; is this the same behavior you see with the example file?

No, it is not. The example file works both with and without `verbatim`. Do you memoize other stuff in your document?

Could you send me the full document over email? Perhaps you have found an example of `verbatim` being harmful.

> In any case, the verbatim section seems a fine place to hint this.

Ok!

**Comment by JasonGross** [comment]
> No, it is not. The example file works both with and without `verbatim`. Do you memoize other stuff in your document?

No.  But I've tracked it down to sticking a `\todo` in in a float with `\let\marginpar\marginnote`:
```latex
\documentclass{article}
\usepackage[accepted]{icml2024}
\usepackage[textsize=tiny]{todonotes}
\usepackage{marginnote}
\let\marginpar\marginnote% to allow \todo in floats
\usepackage[extract=python,verbatim]{memoize}
\icmltitlerunning{}
\begin{document}
\twocolumn[
\icmltitle{}
\begin{icmlauthorlist}
\icmlauthor{Jason Gross}{jg}
\end{icmlauthorlist}
\icmlaffiliation{jg}{\todo{Company Name, Location, Country}}
\vskip 0.3in
]
\printAffiliationsAndNotice{\icmlEqualContribution}
\end{document}
```
(Presumably there's an even smaller example without the icml style.)

**Comment by sasozivanovic** [comment]
Indeed, any document with `\todo` will fail if `verbatim` is in effect. And it is easy to understand why (I don't know how I didn't think of this earlier): `\todo` invokes a `tikzpicture` environment, which Memoize tries to grab verbatim.  However, as the environment was executed from a macro, `\end{tikzpicture}` was already tokenized with standard category codes, so Memoize (actually, CollArgs) will fail to find a verbatim `\end{tikzpicture}`.

In a nutshell, we have found the killer argument against the global `verbatim`: embedded (attempts at) memoization.

FYI, without `verbatim`, Memoize attempts to memoize the `tikzpicture` within `\todo`, but aborts memoization upon encountering `remember picture`.

I guess it's ok to close this again? Please write if you encounter some other issues!

**Comment by JasonGross** [comment]
> I guess it's ok to close this again?

Sure, though I'd still be interested in an obeynewlines or obeyspaces mode that isn't so strong as verbatim, but still handles newline-sensitive code.  Alternatively, is there a setting that allows me to customize the code that gets run at the beginning of memoization, so that I could, e.g., `\everymemoize{\obeyspaces\obeylines}` or `\apptocmd\memoize{\obeyspaces\obeylines}` or something to get the mode I want?

**Comment by sasozivanovic** [comment]
I'm afraid that setting the catcode mode is more complicated than that.  The beginning of memoization (which can be customized through hook `at begin memoization`, as suggested by @cfr42) occurs too late for this change. At that point, the memoized code is already collected, and is in the process of retokenizing via `\scantokens` (under the original catcode regime). To set the catcode mode for collecting the to-be-memoized code, one would need to customize CollArgs — and I have not really foreseen such customization :(

My question is, do we have a piece of code which resists memoization with verbatim but would presumably work in obeynewlines and/or obeyspaces mode?
